src/PolyType/Abstractions/IEnumTypeShape.cs

```diff
@@ -1,26 +1,64 @@
-﻿namespace PolyType.Abstractions;
+﻿using System.Reflection;
+
+namespace PolyType.Abstractions;
 
 /// <summary>
 /// Provides a strongly typed shape model for a .NET enum.
 /// </summary>
-public interface IEnumTypeShape : ITypeShape
+public abstract class IEnumTypeShape(ITypeShapeProvider provider) : ITypeShape
 {
     /// <summary>
     /// Gets the shape of the underlying type used to represent the enum.
     /// </summary>
-    ITypeShape UnderlyingType { get; }
+    public ITypeShape UnderlyingType => UnderlyingTypeNonGeneric;
+
+    /// <inheritdoc/>
+    public abstract Type Type { get; }
+
+    /// <inheritdoc/>
+    public TypeShapeKind Kind => TypeShapeKind.Enum;
+
+    /// <inheritdoc/>
+    public ITypeShapeProvider Provider => provider;
+
+    /// <inheritdoc/>
+    public virtual ICustomAttributeProvider? AttributeProvider => Type;
+
+    /// <inheritdoc cref="UnderlyingType"/>
+    protected abstract ITypeShape UnderlyingTypeNonGeneric { get; }
+
+    /// <inheritdoc/>
+    public abstract object? Accept(TypeShapeVisitor visitor, object? state = null);
+
+    /// <inheritdoc/>
+    public abstract ITypeShape? GetAssociatedTypeShape(Type associatedType);
+
+    /// <inheritdoc/>
+    public abstract object? Invoke(ITypeShapeFunc func, object? state = null);
 }
 
 /// <summary>
 /// Provides a strongly typed shape model for a .NET enum.
 /// </summary>
 /// <typeparam name="TEnum">The type of .NET enum.</typeparam>
 /// <typeparam name="TUnderlying">The underlying type used to represent the enum.</typeparam>
-public interface IEnumTypeShape<TEnum, TUnderlying> : ITypeShape<TEnum>, IEnumTypeShape
+public abstract class IEnumTypeShape<TEnum, TUnderlying>(ITypeShapeProvider provider) : IEnumTypeShape(provider), ITypeShape<TEnum>
     where TEnum : struct, Enum
 {
     /// <summary>
     /// Gets the shape of the underlying type used to represent the enum.
     /// </summary>
-    new ITypeShape<TUnderlying> UnderlyingType { get; }
+    public new virtual ITypeShape<TUnderlying> UnderlyingType => Provider.Resolve<TUnderlying>();
```

Comment:
I would rather this be determined by the deriving type as opposed to hardcoding it to the provider.

Reply:
It's `virtual`. And in some cases there are multiple reflection-based implementations that all do the same thing, so defining it here seems both guaranteed to be correct and in most cases optimal. 
When defining abstract types, I prefer virtual when reasonable and correct behavior can be assumed, yet a derived type might optimize it (as the sourcegen implementation does).

If you feel strongly, sure, I can make this abstract instead. It just feels like more work now, so that future implementations will have to do more work to get off the ground.

Reply:
The default implementation isn't necessary in the case of the source generator. In addition, there are certain corner cases (thinking about union case shapes) where you actually need to return a different shape for the same type. So if this only really applies to the reflection provider, we might as well be intentional and keep it abstract.

Reply:
Fixed.